### PR TITLE
Input plugin for MongoStat

### PR DESCRIPTION
**Pull request by sugr4**
@Microsoft/omsagent-devs 

 $ mongostat --all
insert query update delete getmore command flushes mapped  vsize   res non-mapped faults qr|qw ar|aw netIn netOut conn     time
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:17
    *0    *0     *0     *0       0     3|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  212b    20k    2 16:16:18
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:19
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:20
    *0    *0     *0     *0       0     3|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  212b    20k    2 16:16:21
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:22
    *0    *0     *0     *0       0     3|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  212b    20k    2 16:16:23
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:24
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:25
    *0    *0     *0     *0       0     3|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  212b    20k    2 16:16:26
insert query update delete getmore command flushes mapped  vsize   res non-mapped faults qr|qw ar|aw netIn netOut conn     time
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:27
    *0    *0     *0     *0       0     3|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  212b    20k    2 16:16:28
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:29
    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 25.0M     278.0M      0   0|0   0|0  158b    20k    2 16:16:30

### CODE REVIEW

File: source/code/plugins/mongostat_lib.rb [hunk 1]
```diff
@@ -0,0 +1,164 @@
+module MongoStatModule
+
+  class MongoStat
+    require_relative 'oms_common'
+
+    COUNTER_MAP = {
+       "vsize"      => "Virtual memory Process Usage (MB)",
```
Comment (agup006) [inline source/code/plugins/mongostat_lib.rb]: nitpick - memory to Memory

File: source/code/plugins/mongostat_lib.rb [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+       "lw"         => "% Write Lock Acquisition Time",
+       "lrt"        => "Avg. Read Lock Acquisition Time (ms)",
+       "lwt"        => "Avg. Write Lock Acquisition Time (ms)",
+       "locked"     => "% Time Global Write Lock",
+       "dirty"      => "% WiredTiger Diry Byte Cache",
+       "used"       => "% WiredTiger Cache in Use",
+       "non-mapped" => "Total Virtual Memory (MB)"
+    }
+
+    def transform_data(record)
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: can you give the format of the record?

File: test/code/plugins/mongostat_lib_test.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    set_fields = "insert query update delete getmore command flushes mapped  vsize   res faults qr|qw ar|aw netIn netOut conn     time"
+    send_values = "    *0    *0     *0     *0       0     2|0       0  80.0M 358.0M 41.0M      0   0|0   0|0  133b    10k    1 12:29:52"
+
+    @@mongostat_lib.transform_and_wrap(set_fields)
+    returned_record = @@mongostat_lib.transform_and_wrap(send_values)
+    expected_record = {
+             "DataType"=>"LINUX_PERF_BLOB",
+             "IPName"=>"LogManagement",
+             "DataItems"=> [{
+                   "ObjectName"=>"MongoDB",
```
Comment (robbiezhang) [inline test/code/plugins/mongostat_lib_test.rb]: missing InstanceName in the object

File: source/code/plugins/mongostat_lib.rb [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      if record.start_with?('insert')
+        @counters = arr
+        @values = []
+      else
+        @values = arr
+      end
+
+      if @values and @values.length > 0
+        rec = Hash[@counters.zip @values]
+        @counters.each do |counter|
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: not quite understand this. when @values.length >0, @counters is empty.

File: source/code/plugins/mongostat_lib.rb [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            value = rec[counter]
+            if value == 'nil'
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: value.nil?

File: source/code/plugins/mongostat_lib.rb [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+              rec["qr"] = qr
+              rec["qw"] = qw
+              rec.delete(counter)
+            when "lr|lw"
+              lr, lw = value.split("|")
+              rec["lr"] = lr
+              rec["lw"] = lw
+              rec.delete(counter)
+            when "lrt|lwt"
+              qr, qw = value.split("|")
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: lrt, lwt

File: source/code/plugins/mongostat_lib.rb [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        end       
+ 
+        dataitems = {}       
+        dataitems["Timestamp"] = OMS::Common.format_time(Time.now.to_f)
+        dataitems["Host"] = OMS::Common.get_hostname
+        dataitems["ObjectName"] = "MongoDB"
+        collections = []
+
+        rec.each { |k,v|
+          counter_pair = {"CounterName" => COUNTER_MAP[k], "Value" => v}
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: check has_key before creating the counter_pair

File: source/code/plugins/mongostat_lib.rb [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        }
+        dataitems["Collections"] = collections       
+                  
+        return dataitems
+      end
+    end
+
+    def get_mongostat_version
+      begin
+        version = (%x(mongostat --version)).match(/(\d+)\.(\d+)\.(\d+)/)
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: better to use the named regex

File: source/code/plugins/mongostat_lib.rb [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        num*(1024 ** 3)
+      when "T"
+        num*(1024 ** 4)
+      else
+        num
+      end
+    end
+
+   def transform_and_wrap(record)
+     return {} if record.to_s == ''
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: return nil if record.to_s.empty?

File: source/code/plugins/mongostat_lib.rb [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+     data_items = transform_data(record)
+     if (!data_items.nil? and data_items.size>0)
+       wrapper = {
+          "DataType"=>"LINUX_PERF_BLOB",
+          "IPName"=>"LogManagement",
+          "DataItems"=>[data_items]
+        }
+       return wrapper
+     else
+       return {} unless @values.empty?
```
Comment (robbiezhang) [inline source/code/plugins/mongostat_lib.rb]: return nil

**Comment by sugr4** [comment]
Added comments, changed method tranform_data in mongostat_lib

**Comment by sugr4** [comment]
Changed mongostat.conf to include host, port and run_interval

**Comment by robbiezhang** [comment]
Signoff

**Pull request merged**